### ISSUE TITLE
Translated slugs between locales - linked translations via a reference property. Website with localized URL.

### ISSUE THREAD

**Issue by CarloBu** [sveltia/sveltia-cms#137]
Hello, I like sveltia cms a lot, in my opinion, sveltia cms is currently the best open source cms in the market. 
I'm trying to find a solution for how to make localized markdown filenames work in sveltia cms. The post becomes quite long so I will structure it as readable as possible. 

**Description**
I am currently working on localizing a website, I have done localization with the Astrolicious i18n tool, now I'm trying to work with Sveltia/Decap CMS. Despite extensive searches, I have not found a way to utilize translated filenames for the CMS content. I'm hoping to determine if it's feasible to make it work, I'm looking for any hacky solution as injecting a custom JavaScript into the CMS widget that allows searching for localized posts using translated filenames.

**Current Implementation**
For testing, I utilized the demo repository from the Astrolicious i18n tool and adapted it to my needs. Here are the relevant links:

GitHub Repository: [astrolicious-i18n-demo](https://github.com/CarloBu/astrolicious-i18n-demo)
Live Demo: [astrolicious-18n-demo.netlify.app](https://astrolicious-18n-demo.netlify.app/)

Currently, the testing CMS setup is as follows:

```
config.yaml

collections:
  - name: 'posts'
    label: 'Posts'
    folder: 'src/content/posts'
    create: true
    delete: true
    i18n: true
    format: yml
    extension: yaml
    editor:
      preview: false
    fields:
      - label: 'Default Locale Version'
        name: 'defaultLocaleVersion'
        widget: compute
        value: en/{{fields.title}}
      - label: 'Title'
        name: 'title'
        widget: 'string'
        i18n: true
```
**Objective**
My final goal is to create localized URLs based on slugs taken from Markdown filenames. For instance:

English URL: website/blog/english-post/
French URL: website/fr/le-blog/poste-francaise/
The i18n tool locates the localized YAML file by reading the defaultLocaleVersion, which points to the corresponding default locale file (en/english-post). That way i18n is able to generate localized routes.

**Question**
Is it possible to use some custom widget to enable sveltia cms locating for posts with localized filenames based on this reference?

I've managed to create a post with a default locale reference, but I'm still unable to create localized filenames directly in the CMS and "match" localized posts

Any insights or suggestions on how to implement this feature, or just tell that localized filenames are not possible, would be greatly appreciated.

**Comment by kyoshino** [comment]
> Hello, I like sveltia cms a lot, in my opinion, sveltia cms is currently the best open source cms in the market.

Glad to hear that! There are many things to do before general availability, but I love to see people enjoying the product 💯 

I’m aware of the i18n issue that corresponds to these Netlify/Decap CMS issues:

- https://github.com/decaporg/decap-cms/issues/5493
- https://github.com/decaporg/decap-cms/issues/6600

Well, first-class i18n support is a selling point of Sveltia CMS. I’ll try to find a solution for this!

**Comment by CarloBu** [comment]
Awesome, wow, thank you for your quick response.
Yes, I read these issues while I was researching the solution. I understand that the solution is not so easy as it sounds. I find out that while solving the localized references at frontend :) 
firstly there is a task to generate defaultLocaleVersion. Right now the nw compute widget generates only the first locale item, the rest locales leave with the empty string.
then there is a task for cms to read all files with defaultLocaleVersion and match localized md files with default locale md file

**Comment by kyoshino** [comment]
I think the solution is not complicated, actually.

In Hugo, it’s possible to link localized files using the [`translationKey`](https://gohugo.io/content-management/multilingual/#bypassing-default-linking) property. And yes, there is an issue for Netlify/Decap CMS for supporting:

- https://github.com/decaporg/decap-cms/issues/4645

I’m going to implement it:

1. Allow developers to use the `localize` flag in the configuration file’s [`slug`](https://decapcms.org/docs/configuration-options/#slug) template, like `{{title | localize}}`, in the same way as [summary string transformations](https://decapcms.org/docs/summary-strings/).
2. When the flag is set, the CMS will use the localized value for the file names, while adding the extra `translationKey` property to the files that will hold the default locale’s slug.
3. When the CMS loads the files, use the `translationKey` property to link all the localized files.

**Comment by CarloBu** [comment]
I've encountered an issue with the way filenames are generated from titles, which seems to lack proper sanitization for URL use. For instance, the title "La Cédille En Français" results in the filename la-cédille-en-français.yaml. This filename includes special characters that are undesirable in URLs.

config.yaml

```
- name: test
  label: test
  folder: src/content/test
  create: true
  i18n: true
  format: yml
  extension: yaml
  slug: '{{title}}'
  slug_length: 30
  editor:
    preview: false
  fields:
    - label: Default Locale Version
      name: defaultLocaleVersion
      widget: compute
      value: fr/{{fields.title}}
      i18n: duplicate
    - label: Title
      name: title
      widget: string
      i18n: true
```
      
**Feature Request:**
It would be beneficial to include a slugify option in the compute widget like so:

**config.yaml**
`value: fr/{{fields.title | slugify}}`

This would allow the key value to be generated as a slugified and sanitized URL segment.

**Going further:**
Additionally, it might be valuable to allow developers to implement custom JavaScript functions to define their own widget logic. For example:

javascript
```
function stringify(title) {
  return title.toLowerCase()
              .normalize('NFD').replace(/[\u0300-\u036f]/g, '')
              .replace(/[^a-z0-9 -]/g, '')
              .replace(/\s+/g, '-')
              .replace(/-+/g, '-')
              .replace(/^-+|-+$/g, '');
}
```
This function could be utilized in the configuration like this:

**config.yaml**
```
fields:
  - label: Default Locale Version
    name: defaultLocaleVersion
    widget: javascript
    value: fr/{**{stringify(fields.title)**}}
    i18n: duplicate
  - label: Title
    name: title
    widget: string
    i18n: true
```
Implementing such a feature would enhance flexibility without posing a security risk, as I understand the script execution would be confined to authenticated user contexts. If I'm wrong, please excuse me.

**Comment by kyoshino** [comment]
You can use the [`slug-type`](https://decapcms.org/docs/configuration-options/#slug-type) option 🙂

**Comment by CarloBu** [comment]
Yes! it works and filenames are sanitized now. Somehow it slipped through my eyes :)
thank you!!!
but I still have hard time with
```
  fields:
    - label: Default Locale Version
      name: defaultLocaleVersion
      widget: compute
      value: fr/{{fields.title}}
      i18n: duplicate
```
I get this:
/fr/bibliotheque-publique.yaml
defaultLocaleVersion: fr/Bibliothèque publique
category: Bibliothèque publique
order: 1

but for my i18n setup I need to get this:
/fr/bibliotheque-publique.yaml
defaultLocaleVersion: fr/bibliotheque-publique
category: Bibliothèque publique
order: 1

Is there some option to pass the slug to the defaultLocaleVersion field with added suffix? to be  fr/bibliotheque-publique?

**Comment by kyoshino** [comment]
I think my `translationKey` implementation is ready, so you don’t need the custom `defaultLocaleVersion` stuff. I haven’t added any automated tests, but the change is minimum; it shouldn’t break existing features. 😆

**Comment by kyoshino** [comment]
When you customize the slug like this:

```yaml
slug: '{{title | localize}}'
```

Then file names will be localized, and `translationKey` will be automatically added:

- `en/government-of-canada.yaml`
  ```yaml
  title: Government of Canada
  translationKey: government-of-canada
  ```
- `fr/gouvernement-du-canada.yaml`
  ```yaml
  title: Gouvernement du Canada
  translationKey: government-of-canada
  ```

**Comment by CarloBu** [comment]
omg! Thank you very much!
can't wait to try it

**Comment by kyoshino** [comment]
Just shipped [v0.25.0](https://github.com/sveltia/sveltia-cms/releases/tag/v0.25.0) with the [localized slug support](https://github.com/sveltia/sveltia-cms?tab=readme-ov-file#localizing-entry-slugs) 🎉

**Comment by CarloBu** [comment]
super cool! thank you, kyoshino!
I'm trying it out

**Comment by CarloBu** [comment]
I tried and it is awesome!
I'm trying to work with astrolicious i18n which use translationKey with locale/+slug
by example:
translationKey: fr/my-trip-to-new-york

is there some way to add to my custom key the translationKey with custom string? 
by example:
translationKey: my-trip-to-new-york
+
defaultLocaleVersion: fr/my-trip-to-new-york


collections:
  - name: posts
    label: Blog posts
    create: true
    folder: data/posts/
    slug: '{{title | localize}}'
    format: yaml
    i18n: true
    fields:
      - name: title
        label: Title
        widget: string
        i18n: true
      - name: defaultLocaleVersion
        label: DefaultLocaleVersion
        widget: compute
        value: fr/{{translationKey}}

**Comment by kyoshino** [comment]
I’m playing with your demo project. I think `defaultLocaleVersion` can be replaced `translationKey`, but it seems the locale prefix like `fr/` or `en/` has to be added to frontmatter anyway. Okay, I’ll add a new option to make the key and value customizable.

**Comment by CarloBu** [comment]
astrolicious i18n plugin behind the scenes changes the schema and the frontmatter expects locale/slug, so it's not possible without forking repo.
I don't know what is easier to modify - to add translationKey into the compute widget or give the option in the i18n parameter

but either way it works perfectly:
```
i18n:
  structure: multiple_folders
  locales: [en, fr]
  canonical_slug:
    key: defaultLocaleVersion
    value: 'en/{{slug}}'
```
I understand that is quite a niche issue. :/

**Comment by kyoshino** [comment]
No worries! `translationKey` is specific to Hugo. Other frameworks may have different configurations, so I think it has to be customizable anyway.

**Comment by CarloBu** [comment]
That is very nice of you. I will be able to add sveltia cms to my i18n blog template.
it will be Astro+i18n+SveltiaCMS+Cloudflare stack for ultimate SSG i18n workflow

**Comment by kyoshino** [comment]
Just shipped [v0.26.0](https://github.com/sveltia/sveltia-cms/releases/tag/v0.26.0) with [README](https://github.com/sveltia/sveltia-cms?tab=readme-ov-file#localizing-entry-slugs) updated!

**Comment by CarloBu** [comment]
That's super awesome! Can't believe that you just solved my biggest issue and now astro full localization works with sveltia cms out of the box! I'm going to check if it's working fine with astrolicious i18n.

**Comment by CarloBu** [comment]
Hey, Kyoshino, uh, it's me again,
sorry for this, but I found a bug when the user duplicated the markdown file via sveltiaCMS user interface, the canonical_slug doesn't change. It happens only when using the new canonical_slug. :/
also about UX, when user click on the dublicate button everything stays the same, so at first it's not clear if the button worked, the suggestion is to add suffix to the title field "-COPY" to inform that this is clearly copy and not the original.
by example:
![image](https://github.com/sveltia/sveltia-cms/assets/145860771/ac0688e1-9c5a-4297-8ef7-8f1396fd764b)


https://github.com/sveltia/sveltia-cms/assets/145860771/19146be3-5864-4c09-905b-5a7b175eecf3

**Comment by kyoshino** [comment]
Thanks for finding bugs! Will fix them shortly. There is supposed to be a toast notification after duplicating an entry. Not sure why it’s gone 😅

**Comment by kyoshino** [comment]
Just shipped [v0.26.3](https://github.com/sveltia/sveltia-cms/releases/tag/v0.26.3) to solve the issues!

**Comment by CarloBu** [comment]
now canonical_slug is working at duplication. 
And now I see the first time toaster, nice! 👍

**Comment by steelcj** [comment]
Hi @kyoshino, coming back to this closed issue because the `canonical_slug`/`translationKey` mechanism you built for cross-locale relations is remarkably similar to a solution I'd independently implemented using UUIDs, for a tool I'm building for local, language-based archives with similar goals. I was wrestling with backward compatibility with Dublin Core and suddenly realized that `dc:relation` was infinitely powerful, and that your method and mine had converged on the same underlying pattern.

The underlying pattern is a stable, locale-neutral key (`translationKey`/`UUID` ) that lets a Relation field resolve to "the same thing, expressed differently," independent of where each expression happens to live.

i18n / locale / human written language is one axis of "expressed differently," but it's not the only one. In Functional Requirements for Bibliographic Records (FRBR) terms, the library/archival model for this, a Work is the abstract thing, for example the concept behind an About page or section; each locale version is one Expression of that idea.

Other variants of this relationship are exactly the same shape: a plain-language rewrite, an audio-described version, an Easy Read edition, sign-language images or video, each can be a distinct Expression of the same Work, just as much as a French or English translation is an Expression of the same idea.

Right now `canonical_slug` solves this for the locale axis specifically. But nothing about the mechanism is locale-specific under the hood — it's just "entries that share a stable key are Expressions of one Work." If that key could be configured (or a second relation axis supported) for non-locale groupings — modality, readability tier, format — the same Backlinks/Relation UI you already have would let editors navigate "show me every easy-to-read version of this page" as naturally as they navigate "show me every language of this page" today.

That feels like a meaningfully addition to WCAG and i18n alone: it is the difference between accessibility (and other relationships) being one field on an entry and being a first-class relationship in the content model, using infrastructure you've already built. In other words, the same mechanism can be used to infinitely extend the relationships an Expression of a Work can have.

I figured the concept was worth raising even on a closed issue, since the locale solution already has the shape of the more general one.

BTW, I run a fairly strict evaluation framework for accessible, sustainable, and ethical technology, and Sveltia CMS has consistently rated at the top of it and I have adopted it as the primary publishing vector for my archive tooling. Awesome work my friend, thank you for building such a wonderful tool.

Best, Chris
